typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,726 @@
+These are optics libraries.
+Unfortunately the first one doesn't have type bindings.
+The second one support TypeScript and is a partial porting of Scala monocle.
+And the last one is the newest and TypeScript oriented optics library.
+
+### High Order functions and Currying
+
+A lot of languages support High Order functions and TypeScript is not an exception.
+You can use functions as an argument and also return them.
+
+```typescript
+const filterArray = (fn: (elem: string) => boolean, arr: string[]): string[] => {
+  const newArray: string[] = [];
+  for (let i = 0; i < arr.length; i++) {
+    if (fn(arr[i])) {
+      newArray.push(arr[i]);
+    }
+  }
+  return newArray; // or just simple return arr.filter(fn);
+};
+
+const f = (x: string): ((s: number) => boolean) => {
+  return (s: number) => {
+    return s.toString() === x;
+  };
+};
+
+f("5")(5);
+```
+
+As you may see in the last example, we can call function `f` and return function from it, applying in for the new argument.
+So, yes, we can implement some king of currying using it.
+
+```typescript
+type addT = (_: number) => (_: number) => number;
+const add: addT = (l) => (r) => l + r;
+
+type foldArrT = (f: (_: number) => (_: number) => number) => (_: number) => (_: number[]) => number;
+const foldArr: foldArrT = (f) => (z) => (arr) => {
+  let m = z;
+  arr.forEach((elem) => {
+    m = f(m)(elem);
+  });
+  return m;
+};
+
+foldArr(add)(0)([1, 2, 3, 4, 5]);
+```
+
+Unfortunately syntax is weird as you need to always use named parameters.
+But it is still a possible functionality, and you can simply use it in your code.
+
+### Polymorphism
+
+There are different polymorphisms, but we will say only about thee of them: Parametric, Ad hoc, and Row.
+Let's start from the first on.
+Parametric polymorphism allows us to write abstract function or data which will not depend on its type.
+In TypeScript, it is called generics.
+
+```typescript
+const arrLength: <T>(_: T[]) => number = (arr) => arr.length;
+
+arrLength<number>([1, 2, 3]);
+arrLength<string>(["1", "2", "3"]);
+
+
+type NonEmpty<T> = {
+  head: T;
+  tail: T[];
+};
+```
+
+Ad hoc polymorphism allows you to implement abstract functions, which logic will be different with different types.
+
+```typescript
+interface Eq<T> {
+  equal: (f: T, s: T) => boolean;
+}
+
+class IntEq implements Eq<number> {
+  equal(f: number, s: number) {
+    return f === s;
+  }
+}
+
+class IntArrEq implements Eq<number[]> {
+  equal(f: number[], s: number[]) {
+    return f.filter(item => s.indexOf(item) < 0).length === 0;
+  }
+}
+
+const lookup = <T, K extends Eq<T>, V>(cmp: K, k: T, mp: [T, V][]): V | undefined => {
+  let result: V | undefined;
+  mp.forEach(([kk, v]) => {
+    if (cmp.equal(k, kk)) {
+      result = v;
+    }
+  });
```

Comment:
In this particular case, is that true that you could go without the `result` variable and just do `return v` when the value is found?

Reply:
`return` in `foreach` make nothing, just didn't want to use `for`, but I will try to make a better example

Reply:
Ah right, fair enough.